crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -155,31 +157,46 @@ impl Engine for WasmtimeEngine {
             .into_error_code()
     }
 
-    fn precompile(&self, layers: &[WasmLayer]) -> Result<Vec<Option<Vec<u8>>>> {
-        let mut compiled_layers = Vec::<Option<Vec<u8>>>::with_capacity(layers.len());
-
-        for layer in layers {
-            if PRECOMPILER.detect_precompiled(&layer.layer).is_some() {
-                log::info!("Already precompiled");
-                compiled_layers.push(None);
-                continue;
-            }
-
-            use WasmBinaryType::*;
-
-            let compiled_layer = match WasmBinaryType::from_bytes(&layer.layer) {
-                Some(Module) => PRECOMPILER.precompile_module(&layer.layer)?,
-                Some(Component) => PRECOMPILER.precompile_component(&layer.layer)?,
-                None => {
-                    log::warn!("Unknow WASM binary type");
+    #[allow(clippy::manual_async_fn)]
+    fn precompile(
+        &self,
+        layers: &[WasmLayer],
+    ) -> impl Future<Output = Result<Vec<PrecompiledLayer>>> + Send {
+        async move {
+            let mut compiled_layers = Vec::<PrecompiledLayer>::new();
+
+            for layer in layers {
+                if PRECOMPILER.detect_precompiled(&layer.layer).is_some() {
+                    log::info!("Already precompiled");
                     continue;
                 }
-            };
 
-            compiled_layers.push(Some(compiled_layer));
-        }
+                use WasmBinaryType::*;
+
+                let compiled_layer = match WasmBinaryType::from_bytes(&layer.layer) {
+                    Some(Module) => PRECOMPILER.precompile_module(&layer.layer)?,
+                    Some(Component) => PRECOMPILER.precompile_component(&layer.layer)?,
+                    None => {
+                        log::warn!("Unknow WASM binary type");
+                        continue;
+                    }
+                };
+
+                let parent_digest = layer.config.digest().to_string();
```

Comment:
is parents expected to never be empty?

Reply:
At the moment, i think that should be the assumption. Otherwise an empty parents set would imply a new layer that has no relation to the other layers in the image. I'll add a check in the loop above to ensure this precondition for now.

Reply:
Can you also specify this in the inline doc comments on the `PrecompiledLayer` structure? Ill leave a comment there too

Reply:
Oh yeah, good catch. Will update the comment. I'd be wary of the `self if none` approach as that would introduce a cycle we'd have to manage when loading.